README.md
# Facility Registry API Server

A RESTful API that serves basic facility information, designed for simple integration with other applications that use facilities.

[![Build Status](https://travis-ci.org/SEL-Columbia/Revisit-Server.png)](https://travis-ci.org/SEL-Columbia/Revisit-Server)

Although we are not limiting the data to health sector facilities, the service will conform to the [FRED API specification](http://facilityregistry.org/) as closely as possible, with several extensions.

For details on setting up a local dev enviroment, [see the wiki](https://github.com/SEL-Columbia/Revisit-Server/wiki/Setting-Up-a-Local-Environment).

# Notes

## Extra Properties
* Number of times visited
* Sector
* Type
* Does the facility have power?

## Dependencies
* Restify: REST API framework for node.js
* Mongo: Document store, a lot of node integration options
* Mongoose: ODM for mongo, enforces facility format
* jshashes: hashing basic auth passwords
* mkdirp: simple recursive directory creation for photos
* bunyan: logger
* mocha: For BDD testing
* supertest: for testing requests
* should: readable asserts 

## Possible Alternatives
* Save: DB interaction layer, allows us to swap mongodb without much code change
* Express: More involved stack then restify, has restify extension as well
* Mongojs: Exposes mongoapi directly


## Deployment
* Ruby based deployment via [Capistrano](http://capistranorb.com). For details on deployment, [see the  wiki](https://github.com/SEL-Columbia/Revisit-Server/wiki/Deployment).


CHANGES.txt
This file contains differences between our api and the FRED specs minus the geo
queries which ofcourse are not part of the FRED spec altogether 

Facility Add:

    Currently wipes the following fields from the post:

    var badKeys = [
                '_id',
                'uuid',
                'url',
                'createdAt',
                'id',
                'href',
                '__v',
                'updatedAt'
               ];

    uuid however is saved and turned into an objectID internally. 
    The user however will not see this, in the end if they post a valid uuid,
    they will see their uuid in the response. uuid MUST be a 24 char hex or a 12
    byte string.

    It also requires you to provide a sector and a name for the facility.

    FRED allows id to be passed in (in any format to either id/uuid not sure)
    FRED does not reject any field
    FRED implies name is required but does not enforce it.
    FRED does not enforce sector.


Facility Update:

    Currently wipes the following fields from the request:

    var badKeys = [
                '_id',
                'uuid',
                'url',
                'createdAt',
                'id',
                'href',
                '__v',
                'updatedAt'
               ];

    FRED does not allow "id" "createdAt" and "updatedAt"

    Theres no reason to allow for updating id/uuid.

Facilities GET:

    Like FRED we respond with a facility object however we do not allow for
    limit = off to actually be off. limit = off is now limit = 1000.

    To give the user some idea of how many results there in the db the following
    three fields were added to the faciliies json reply:

        - total: total #results in db for this query
        - length: current number returned
        - offset: current offset in total available   

    FRED does none of this, instead provides limit=off to actually return
    everything.

    TODO: Jon update this with pagination details!

Facility Bulk:

    We provide a way to do bulk uploads from a file (and as a json array but
    that endpoint is currently not exposed)

    It takes in a facility object like the one outputed by the main get
    endpoint. 

    It does validation against the site model and does not do the validation
    that update/add do. I.e you can provide all the fields in the badkey list
    but you cannot omit required fields (sector and name). This makes sense
    since fields like createdAt are probably useful to keep around in this case.

    _id is removed, just incase.

    However if you provide a uuid, like add this field is saved as _id and shows
    in the response as a uuid. It once again expects the format to be a 24 char
    string.


    Bulk endpoint never 404/400's unless something major happened. A report is
    sent as a response explaining what was and wasnt inserted. Add ?debug to get
    a detailed error report if needed.


Facility GET:
    
    Three new query paramters were added. hist, rollback and revert.
    They return history, updated the facility with a previous version and revert
    to a previous facility respectively. 

    If model is inserted outside of the api OR through the bulk upload enpoint
    version 0 is not recorded. Version 0 is created on update to the model (so
    version 0 == first update in this case) or through any of the three hist
    queries (Version 0 === actual version 0 in this case, generated on the fly)



routes.js
var SiteModel = require('./models/site').SiteModel,
    restify = require('restify'),
    fs = require('fs'),
    mkdirp = require('mkdirp'),
    prePath = '/api/v1';

exports.sites = function(req, res, next) {
    var sites = SiteModel.find(function(err, sites) {
        if (err) console.log(err);
        res.send(sites);
    });
}

exports.site = function(req, res, next) {
    var id = req.params.id,
        query = {};

    // check if valid ObjectID
    if (id.match(/^[0-9a-fA-F]{24}$/)) {
        query._id = id;
    } else {
        query.uuid = id;
    }

    console.log(query);
    SiteModel.find(query, function(err, site) {
        console.log(err);
        if (err) return next(new restify.InvalidArgumentError(JSON.stringify(err.errors)));

        if (site) {
            res.send(site)
        } else {
            res.send(404)
        }
    });
}

exports.near = function(req, res, next) {
    // console.log(req.query); return;
    var lat = req.params['lat'],
        lng = req.params['lng'],
        rad = req.params['rad'] || 10,
        units = req.query['units'] || 'mi',
        earthRad = 3959; // miles

    if (units === 'km') {
        earthRad = 6371;
    }

    SiteModel.find({
        "coordinates": {
            "$geoWithin": {
                "$centerSphere": [
                    [lng, lat], rad / earthRad
                ]
            }
        }
    }).exec(function(err, sites) {
        if (err) console.log(err);
        res.send(sites);
    });
};

exports.within = function(req, res, next) {
    console.log(req.query.sector);
    var swlat = req.params['swlat'],
        swlng = req.params['swlng'],
        nelat = req.params['nelat'],
        nelng = req.params['nelng'],
        sector = req.query['sector'] || null,
        coordinatesQuery = {
            "coordinates": {
                "$geoWithin": {
                    "$box": [
                        [swlng, swlat],
                        [nelng, nelat]
                    ]
                }
            }
        },
        query = {};

    if (sector) {
        var properties = {
            "properties.sector": sector
        };
        query = {
            "$and": [
                coordinatesQuery,
                properties
            ]
        };
    } else {
        query = coordinatesQuery;
    }

    SiteModel.find(query).exec(function(err, sites) {
        if (err) console.log(err);
        res.send(sites);
    });
};

exports.newSite = function (req, res, next) {
    // TODO: validity checks
    // console.log(req.body);
    var site = new SiteModel(req.body);
    site.save(function(err, site) {
        console.log('save callback...', err, site);
        if (err) {
            console.log('!!!!!!!!! ERROR !!!!!!!!!!', err);
            return next(new restify.InvalidArgumentError(JSON.stringify(err.errors)));
        } else {
            res.send(site);
            console.log(':):):):):) Success :):):):):)', res);
            next();            
        }
    });
}

exports.updateSite = function (req, res, next) {
    var site = req.body,
        id = site['_id'],
        query = {};

    // At the moment, we require an _id here, not uuid since mongoose expects this for the findByIdAndUpdate method
    site.updatedAt = Date();

    SiteModel.findByIdAndUpdate(id, {$set: site}, function (err, site) {
      if (err) return next(new restify.InvalidArgumentError(JSON.stringify(err.errors)));
      res.send(site);
    });
}

exports.deleteSite = function (req, res, next) {
    return next(new restify.RestError({statusCode: 400, restCode: "Not Implemented", message: "Delete method not yet implemented."}));
}

exports.flagSite = function (req, res, next) {
    return next(new restify.RestError({statusCode: 400, restCode: "Not Implemented", message: "Flag method not yet implemented."}));
}

exports.uploadPhoto = function (req, res, next) {
    var siteId = req.params.id || null;

    console.log(req.files.photo);
    
    // if no sideId is included in request, error
    if (!siteId) {
        return next(new restify.MissingParameterError("The required siteId parameter is missing."));
    }

    // make sure the id is associated with a known Site
    SiteModel.findById(siteId, function (err, site) { 
        if (err) return next(new restify.ResourceNotFoundError(JSON.stringify(err)));

        console.log(site);

        // move the uploaded photo from the temp location (path property) to its final location
        fs.readFile(req.files.photo.path, function (err, data) {
    		if (err) {
    			console.log(err);
    		}
            var rootPath = "/home/ubuntu/facrest/public/sites/photos",
                siteDir = siteId,
                filePath = req.files.photo.name,
                fullPath = rootPath + '/' + siteDir + '/' + filePath;

            // create the dir for the site
            mkdirp(rootPath + '/' + siteDir, function (err) {
                if (err) {
                    console.error(err);
                    return next(new restify.InternalError(JSON.stringify(err)));
                } else {
                    fs.writeFile(fullPath, data, function (err) {
                        console.log('writeFile callback');
            			if (err) {
            				console.log('write error: ' + err);
            				return next(new restify.InternalError(JSON.stringify(err)));
            			}

                        var url = 'http://' + req.header('Host') + '/sites/photos/' + siteDir + '/' +  filePath;

                        // check that this photo is new. we'll replace the tmp path with the url
                        var index = site.properties.photoUrls.indexOf('tmp/'+req.files.photo.name);
                        if (index != -1) {
                            site.properties.photoUrls.splice(index, 1, url);
                            // site.properties.photoUrls.push(url);
                        }

                        console.log('site photo url: ' + url);
			
                        site.save(function (err, updatedSite, numberAffected) {
                            if (err) {
                				console.log('save error: ' + err);
                				return next(new restify.InternalError(JSON.stringify(err)));
            				}
                            console.log('site saved, sending response');
				            // no error, send success
                            res.send(updatedSite);
                        });
                    });
                }
            });
        });
    });
}
